### ISSUE TITLE
Domesticating Kubernetes | Practical kubernetes as a home server

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6518]
Domesticating Kubernetes | Practical kubernetes as a home server<br>
<br>
https://ift.tt/3aScjXH<br>
<br>
Vladimir Akopyan<br>
<br>
<div>
<div>
<div>
<h2>Kubernetes as home server on bare metal in 150 minutes</h2>
<div>
<div>
<div>
<div><a rel="noopener" href="https://blog.quickbird.uk/@ClumsyPilot?source=post_page-----d49c178ebc41----------------------"><img alt="Vladimir Akopyan" width="48" src="https://miro.medium.com/fit/c/96/96/1*xGwsC9BcQBS6uEW2ewP4BQ.jpeg" height="48"></a></div>
</div>
</div>
</div>
<p>This is a guide to run K8S in a home network, and use it as a home server &mdash; run your blog, media library, <a rel="noopener" href="https://www.home-assistant.io/" target="_blank">smart home</a>, pet projects, etc.<br>
The cluster is actually straight-forward to set up, but we, developers are so cuddled, we are forgetting some basic networking and other low-level stuff &mdash; I found the experience educational.</p>
<p>The cluster will serve real workloads &mdash; we will deal with exposing it to the internet, IP assignments in home network, reasonable security, distributed storage and monitoring.<br>
It is aimed at a home network, and does not rely on loadbalancers, SAN&rsquo;s, multiple public IPs or any other fancy infrastructure. I am keeping it as simple (read reliable) as possible &mdash; there are no &lsquo;enterprise&rsquo; bells and whistles.</p>
<p>To proceed, make sure you are comfortable with basic kubernetes concepts, know what&rsquo;s a master node, an agent, a LoadBalancer service, a deployment, ingress, persistent volume, etc.</p>
<p>Let&rsquo;s consider K8S cluster as a layered cake and take a look at each layer</p>
<figure>
<div>
<div>
<div>
<div>
<div><img width="1292" src="https://miro.medium.com/max/60/1*jZEcmmeIgSKJ3ZHFKN7cdw.png?q=20" height="954"></div>
<img src="https://miro.medium.com/max/2584/1*jZEcmmeIgSKJ3ZHFKN7cdw.png" width="1292" height="954">
</div>
</div>
</div>
</div>
</figure>
<p>At the top are the <strong>Applications</strong> that you are writing and/or running &mdash; this is the part that actually delivers value and where <strong>developers</strong> will spend most of their time. This might be your wordpress blog, some API you&rsquo;ve written and your bitcoin trading bot.</p>
<p>Next level down are <strong>Services</strong> for administration and running the applications &mdash; that&rsquo;s your own MySQL database, <a rel="noopener" href="https://www.elastic.co/what-is/elk-stack" target="_blank">ELK Stack</a>, Monitoring, etc. They don&rsquo;t have to run in your cluster &mdash; Amazon/Azure/GCP offer PAAS versions with their managed K8S serviceg. <strong>DevOps</strong> and administrators are spending a lot of their time here.</p>
<p>At the <strong>System-level</strong> we&rsquo;ve got the components that make up a functional cluster&mdash; you can&rsquo;t skip on any of these:</p>
<ul>
<li>software components of K8S (kubelet, API-server, etc.)</li>
<li>storage provider for K8S persistent volumes</li>
<li>authentication provider for kubernetes users</li>
<li>loadbalancing and external ips</li>
</ul>
<p>Smaller managed K8S providers like O<a rel="noopener" href="https://www.ovhcloud.com/en-gb/public-cloud/kubernetes/#https://www.ovhcloud.com/en-gb/public-cloud/prices/" target="_blank">vhCloud</a> and <a rel="noopener" href="https://www.digitalocean.com/products/kubernetes/" target="_blank">DigitalOcean</a> typically operate at this level. You have to configure them if you are bootstrapping your own cluster. System administrators and IT services might be spending majority of their time here.</p>
<p><strong>Infrastructure</strong> layer is self-explanatory &mdash; that&rsquo;s the metal, CPU, RAM, Disk, and physical network.</p>
<p>You might be tempted to get a bunch of Rasberri Pi&rsquo;s, but there are better alternatives.Before we dive into them, consider the following:</p>
<ul>
<li>CPU and RAM get pooled together in a cluster, you can get a solid 20GB ram and 6 cores out of a couple old laptops or other outdated kit lying around.<br>
I would not buy old rack servers &mdash; they are cheap and powerful, but they sound like a stalling turbofan and their power consumption is crazy.</li>
<li>Storage works the other way &mdash; we will install a distributed storage system on our cluster, they (typically) keep 3 copies of data for redundancy. SBCs like the PI are very gimped in this regard, their performance is 10x lower than that of anything with a proper SSD, and reliability is lower.<br>
Mini-pc like the Intel-NUC usually come in two versions &mdash; the &lsquo;fatter&rsquo; one, it has an extra SATA drive slot, which you could use for an HDD.</li>
<li>You want a system with &lsquo;always-on&rsquo; functionality, so that the computer starts itself after power loss, at least for the master-node. It&rsquo;s in BIOS settings of most desktops and SBCs, but most laptops don&rsquo;t have it.</li>
</ul>
<p>Here is my K8S cluster, it fits on a single shelf in the closet:</p>
</div>
</div>
<div>
<div>
<p>All the kit is plugged into a gigabit Ethernet switch. Left to right, these are:</p>
<ul>
<li>Beelink Gemini X45 with J4105 8GB RAM, 128GB SSD and 320 GB HDD, this is the master node.</li>
<li>Intel NUC with 5th Gen i3, 8Gb RAM, 128GB SSD and 320 GB HDD</li>
<li>Linx1010B &mdash; an joke of a windows tablet, Intel Atom, 2GB Ram.</li>
<li>An old Samsung laptop with 3rd gen i5, 6GB Ram and 256GB SSD</li>
<li>An obligatory PI4, 2GB (not pictured)</li>
</ul>
</div>
</div>
<div>
<div>
<p>Looking at the benchmark, Raspberri PI&rsquo;s hardly make any sense:</p>
<ul>
<li>PI4 with 4Gb ram, sd card, case,etc. is about &pound;100. For the same money you can get a no-name Intel-atom mini-pc, and those come with the benefit of x86 arch, real bios and real Sata or m.2 ports.</li>
<li>My Beelink set me back about ~&pound;150 and it&rsquo;s a noticeable upgrade</li>
<li>For &pound;200 you can buy used mini-desktops, like ThinkCentre M700 i5&ndash;6400T. That a major performance improvement, but the device is larger.</li>
<li>At &pound;300 and up you can build a brand-new, compact HTPC system, for example based on ASRock DeskMini A300 and full-power desktop components. Or you can always go with Intel-NUC if space is at a premium.</li>
</ul>
</div>
</div>
<div>
<div>
<p>First and foremost if you want to host any web-services you need to make sure aren&rsquo;t behind <a rel="noopener" href="https://en.wikipedia.org/wiki/Carrier-grade_NAT" target="_blank"><strong>carrier-grade NAT</strong></a>. My provider uses it by default, but I got a static IP for extra &pound;5 a month.</p>
<p>Next, let&rsquo;s assume you have a <strong>DNS registrar</strong>, got yourself the domain <strong>timmy.com.</strong> Unlike in a typical deployment in the cloud, we have only one IP address to play with, so setup records to direct traffic from <strong>timmy.com</strong> and <strong>*.timmy.com</strong> (any subdomain) to your public IP address, so it arrives at your router.</p>
<p>Behind your router, your LAN IPs will be split into three ranges:</p>
<ul>
<li>A range for static IPs assigned to important devices in your home network, it typically starts with your router, i used 192.168.0.1&ndash;255. All computers / nodes in the cluster should be given a static IP.</li>
<li>A range for DHCP assignments, this is for various devices that connect to your network &lsquo;just to use the internet&rsquo;, like your mobile phone. I configured DHCP server in the router to use 192.168.0&ndash;255.</li>
<li>Kubernetes services will have a floating IP addresses of their own, and the actual service might be located on any of the nodes in our cluster, depending on load and the whims of the kubernetes scheduler. We will be using <a rel="noopener" href="https://v0-3-0--metallb.netlify.app/concepts/arp/" target="_blank">MetalLB in ARP mode</a> to achieve this, and the range of 192.168.2.0&ndash;255.</li>
</ul>
<blockquote>
<p>I have changed subnet /netmask of my router to 255.255.240.0. The actual range you use does not matter, you could leave default router subnet and use the &lsquo;higher&rsquo; end IPs of 220&ndash;250 for static IP and load balancing. If you pick a different subnet, an <a rel="noopener" href="https://www.calculator.net/ip-subnet-calculator.html" target="_blank">IP calculator</a> can help.</p>
</blockquote>
<p>Once the traffic arrives at your router, we have to use port-forwarding to direct it to the right place. Traffic for the Kubernetes API server, typically on TCP:6443, must be directed to the master node &mdash; this will enable you to connect to your cluster using Kubectl from the internet.</p>
<blockquote>
<p>In this setup we are only considering a single master node &mdash; if you had several of them for HA, you&rsquo;d have to configure <a rel="noopener" href="https://www.keepalived.org/" target="_blank">keepalived</a> or <a rel="noopener" href="http://www.haproxy.org/" target="_blank">HAproxy</a>, or both.</p>
</blockquote>
</div>
</div>
<div>
<div>
<p>Notice that only services of type LoadBalancer will be given an IP address on your LAN network. All other resources will reside on a VLAN setup with flannel, they can reach each-other but are isolated from the outside world.</p>
<p>Traffic on TCP:80 and 443 must be directed to the <a rel="noopener" href="https://kubernetes.io/docs/concepts/services-networking/ingress/" target="_blank">ingress service</a> using it&rsquo;s IP &mdash; from there it will be routed to the correct application depending on the domain name, and we can host virtually unlimited number of websites that way.</p>
<p>Only HTTP traffic can be routed based on domain name, so if we want to expose a MySQL database, we must port-forward that particular service. If we have two such databases, we have to give them different ports.</p>
<p>To proceed you need to have setup a domain / DNS records, have decided on your IP ranges and have your router / DHCP configured accordingly.</p>
<p>I have chosen Ubuntu Server 20.04 LTS, just because of familiarity and it&rsquo;s ubiquity &mdash; there is even a version for Raspberri PI. In this setup, very little depends on a particular OS. Install it on each node, consider the following:</p>
<ul>
<li>Stick to simple alphanumerics in the hostname of each computer or Kubernetes won&rsquo;t start and you will have to specify a K8S-acceptable name for the node separately.</li>
<li>If you plan to use the same drive for OS and for storing data of <a rel="noopener" href="https://kubernetes.io/docs/concepts/storage/persistent-volumes/" target="_blank">Persistent Volumes</a>, enable LVM &mdash; this will allow you to split the primary disk into an OS and storage partition, and resize them as needed.</li>
<li>Use <code>ssh-import-id-gh &lt;username&gt;</code> if you need to import SSH key from Github after installing OS without internet access.</li>
<li>
<a rel="noopener" href="https://linuxconfig.org/how-to-configure-static-ip-address-on-ubuntu-18-04-bionic-beaver-linux" target="_blank">Configure a static IP address</a> in the <code>/etc/netplan/50-cloud-init.yaml</code> , I prefer to keep DHCP enabled in addition to static IP.</li>
<li>If you are using a laptop, <a rel="noopener" href="https://askubuntu.com/questions/141866/keep-ubuntu-server-running-on-a-laptop-with-the-lid-closed" target="_blank">disable suspend on closing laptop lid</a>. I don&rsquo;t know why a server OS does that!</li>
</ul>
<p>To proceed, make sure all your nodes are setup and you can SSH into all of them.</p>
<p>Kubernetes is like linux &mdash; there are different takes on it, and for a homelab <a rel="noopener" href="https://microk8s.io/" target="_blank">MicroK8S</a> and <a rel="noopener" href="https://k3s.io/" target="_blank">K3S</a> make the most sense as the two simplified distributions.<br>
Pick the most reliable/fastest/whatever machine, and that will be our master-node. Begin installing K8S with it.</p>
<h2>MicroK8s vs K3s</h2>
<p>My experience with MicroK8s has been substantially better &mdash; it is mostly a vanilla K8S packaged into a Snap, if you want to understand what it&rsquo;s doing, you can read the standard configuration files for kubelet, kubeapi server, etcd, etc.</p>
<p>K3S is much stranger &mdash; all components of K8S have been packed into a single binary, and run as a single service/deamon. In my mind there are only three reasons to use K3S:</p>
<ol>
<li>You are not satisfied with etcd and want to use the <a rel="noopener" href="https://rancher.com/docs/k3s/latest/en/installation/datastore/" target="_blank">unique K3S databastore options</a>: either running SQLite, or PostgreSQL/MySQL</li>
<li>You really need to minimise resource overhead of K8S</li>
<li>You want to install Rancher server in the cluster to take advantage of it&rsquo;s great UI and Auth features. You can only install Rancher on <a rel="noopener" href="https://rancher.com/docs/rancher/v2.x/en/installation/k8s-install/" target="_blank">K3S or RKE c</a>lusters.</li>
</ol>
<pre><span>The Rancher management server can only be run on Kubernetes cluster in an infrastructure provider where Kubernetes is installed using K3s or RKE. Use of Rancher on hosted Kubernetes providers, such as EKS, is not supported.</span></pre>
</div>
</div>
<div>
<div>
<h2>K3S configuration &mdash; option 1</h2>
<p>K3S comes with lots of components we want to replace.</p>
<pre><span>curl -sfL <a rel="noopener" href="https://get.k3s.io" target="_blank">https://get.k3s.io</a> | sh -s - --write-kubeconfig-mode 644 <strong>--tls-san</strong> k3s.timmy.com --no-deploy servicelb --no-deploy local-storage</span></pre>
<ul>
<li>
<strong>TLS-SAN</strong> &mdash; this parameter is necessary if you want to be able to access your cluster from the internet through kubectl. It is used to generate a certificate for kube-api server. The certificate will be valid for the LAN IP address of the machine and the domain you specify. This information is encoded in <code>certificate-authority-data</code>in the kubectl config file.</li>
<li>
<strong>ServiceLB</strong> &mdash; it can&rsquo;t create IP addresses independent of a machine, we will be replacing it with MetalLB</li>
<li>
<strong>LocalStorage &mdash;</strong> provisions storage through hostpath. We will be replacing it with proper distributed storage.</li>
</ul>
<p>Once the command is complete, your masternode should be up and running. Retrieve your kubeconfig from<code><strong>/etc/rancher/k3s/k3s.yaml</strong></code> and merge / replace <code>kubeconfig</code> on your personal machine. Replace the <code>server: <a rel="noopener" href="https://127.0.0.1:16443" target="_blank">https://127.0.0.1:16443</a></code> with the domain name of the you spesified above &mdash; for example k3s.timmy.com. Validate that <code>kubectl</code> works form your dev machine and you can get pods, etc.</p>
<p>To add other machines as agents in the cluster, retrieve the token from <code><strong>/var/lib/rancher/k3s/server/node-token</strong></code> on the master node.<br>
You can then get them to join the cluster by running:</p>
<pre><span><strong>curl -sfL https://get.k3s.io | K3S_URL=192.168.1.2:6443 K3S_TOKEN=mynodetoken sh -</strong></span></pre>
<p>Avoid using domain name for connecting agents to the master node &mdash; it will work but any issues with DNS will result in your cluster falling apart. Validate that you have a collection of functional nodes with <code>kubectl get nodes</code></p>
<h2>MicroK8S configuration &mdash; option 2</h2>
<p>MicroK8S comes with a rich CLI tool that allows you to inspect and configure a cluster:</p>
<pre><span>sudo snap install microk8s --classic --channel=1.18/stable<br>microk8s status --wait-ready<br>microk8s kubectl get nodes</span></pre>
<p>To enable access to kube-api server through it&rsquo;s public IP and DNS name, edit <code>/var/snap/microk8s/current/certs/csr.conf.template</code> to include them. It will look something like this:</p>
<pre><span>...<br>[ alt_names ]<br>DNS.1 = kubernetes<br>DNS.2 = kubernetes.default<br>...<br>DNS.6 = timmy.com<br>DNS.7 = microk8s.timmy.com</span><span>IP.1 = 127.0.0.1<br>IP.2 = 10.152.183.1<br>IP.3 = 192.168.1.2<br>...</span></pre>
<p>The `apiserver-kicker` will automatically detect the difference, generate new certificated and restart the apiserver. Unlike K3S, we can have as many domain names as we please.</p>
<p>Retrieve <code>kubeconfig</code> using <code>microk8s config</code> command and merge / replace <code>kubeconfig</code> on your personal/dev machine . Replace the server IP address with it&rsquo;s proper DNS name, or you could have two entries in your <code>kubeconfig</code> &mdash; one for local access, and one for remote.</p>
<h2><strong>Installing MetalLB</strong></h2>
<p>On MicroK8S you install MetalLB by enabling the corresponding addon. SSH into masternode an execute: <code>microk8s enable metallb</code> . It will ask you for an IP range you&rsquo;d like to use.</p>
<p>On K3S you must <a rel="noopener" href="https://metallb.universe.tf/installation/" target="_blank">install MetalLB through kubectl:</a></p>
<pre><span>kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.9.3/manifests/namespace.yaml<br>kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.9.3/manifests/metallb.yaml<br># On first install only<br>kubectl create secret generic -n metallb-system memberlist --from-literal=secretkey="$(openssl rand -base64 128)"</span></pre>
<p>Then you must create a <strong>configmap</strong> in the <code>metallb-system</code> namespace to specify the IP range it can use:</p>
<pre><span>kind: ConfigMap<br>apiVersion: v1<br>metadata:<br> name: config<br> namespace: metallb-system<br>data:<br> config: |<br> address-pools:<br> - name: default<br> protocol: layer2<br> addresses:<br> - 192.168.2.0-192.168.2.254</span></pre>
<p>Verify that MetalLB works by deploying a blank nginx application with service of type <code>LoadBalancer</code></p>
<pre><span>apiVersion: apps/v1<br>kind: Deployment<br>metadata:<br> name: hello<br> labels:<br> app: hello<br>spec:<br> replicas: 3<br> selector:<br> matchLabels:<br> app: hello<br> template:<br> metadata:<br> labels:<br> app: hello<br> spec:<br> containers:<br> - name: nginx<br> image: nginx:1.14.2<br> ports:<br> - containerPort: 80<br>---<br>kind: Service<br>apiVersion: v1<br>metadata:<br> name: hello<br> labels:<br> app: hello<br>spec:<br> ports:<br> - name: http<br> protocol: TCP<br> port: 80<br> targetPort: 80<br> selector:<br> app: hello<br> type: LoadBalancer<br> externalTrafficPolicy: Cluster</span></pre>
<p>It should be assigned a valid LAN IP and be reachable form your dev/personal computer.</p>
<p>Ingress software is not part of the Kubernetes software project, instead <a rel="noopener" href="https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/" target="_blank">Ingress Controllers</a> are third party software that is installed in a cluster and configured by Kubernetes &mdash; like anything else, they run in a pod/container, and needs a service to be reachable form the outside world. Each has it&rsquo;s perks, but they fulfil the same need. For all of them you should:</p>
<ol>
<li>Switch the ingress type from <strong>ClusterIp</strong> to <strong>LoadBalancer</strong> to so that it&rsquo;s assigned an IP address on our LAN</li>
<li>Configure the router to port-forward TCP connections on port 80 (http) and 443 (https) to this address.</li>
<li>Explicitly specify a <code>loadBalancerIP</code> so that you don&rsquo;t have to change router settings if you need to re-install ingress for some reason.</li>
<li>Pick an address near the end of the available range, so that some other service does not occupy it and get in the way &mdash; <code>MetalLb</code>assigns them sequentially.<br>
Snippet from the ingress <code>yaml</code> definition file:</li>
</ol>
<pre><span>kind: Service<br>apiVersion: v1<br>spec:<br> type: LoadBalancer<br> loadBalancerIP: 192.168.2.100</span></pre>
<h2>Nginx on MicroK8S</h2>
<p>Nginx is considered the standard ingress. It&rsquo;s pre-installed on MicroK8S. Edit existing ingress service in accordance with the above, and you are done.</p>
<h2><strong>Traefik on K3S</strong></h2>
<p>There are a couple advantages to using Traefik &mdash; it&rsquo;s comes with a pretty dashboard and unlike nginx it can update configuration without reloading. Additionally, it&rsquo;s smart enough to realise that any service with port 443 or port names https requires https connection (shock!).<br>
The downsides are &mdash; there is less documentation and it&rsquo;s less powerful when it comes to acting as an authentication proxy &mdash; it does not support OAUTH authentication out of the box, and needs an extra component if you want t authenticate with Github, etc.</p>
<p>Traefik comes pre-installed on K3S, but we need to modify it&rsquo;s configuration. Do not modify existing kubernetes resources &mdash; K3S has an annoying add-on-like system, where it will monitor manifests in <code>/var/lib/rancher/k3s/server/manifests/</code> for changes, and deploy them into your cluster. Any changes you make directly to the kubernetes resources will be overwritten.<br>
Instead, edit the <code>traefik.yaml</code> file in the manifests folder. It is basically a helm chart values file. Set the following vlaues, in addition to defaults:</p>
<pre><span>ssl.enabled: "true"<br>ssl.insecureSkipVerify: "true" <br>metrics.prometheus.enabled: "true"<br>metrics.serviceMonitor.enabled: "true" <br>dashboard.enabled: "true"<br>dashboard.serviceType: "LoadBalancer"<br>dashboard.auth.basic.admin: "$apr1$tM.asdgfs$kljkuwd" <br>loadBalancerIP: 192.168.2.100<br>logLevel: "debug"</span></pre>
<ul>
<li>don&rsquo;t use <code>ssl.enforced</code> , it breaks cert-manager&rsquo;s let&rsquo;s encrypt validation</li>
<li>
<code>ssl.insecureSkipVerify</code> will allow Traefik to work with HTTPS backends with untrusted / self signed certificates. This can be useful for Kubernetes dashboard and in-cluster databases. I consider this acceptable in home deployments.</li>
<li>Prometheus options are usefull for monitoring, but we will not cover their use in this tutorial.</li>
<li>
<code>dashboard.enabled</code> enables <a rel="noopener" href="https://docs.traefik.io/operations/dashboard/" target="_blank">Traefik dashboard</a>, which to me is half the value of having it as an ingress controller. Use <code>htpasswd</code>utility to generate SHA hash for <code>dashboard.auth.basic.admin</code> , where admin is the username.</li>
</ul>
<p>Save the resulting file as <code>traefik-customised.yaml</code> and delete the original &mdash; <a rel="noopener" href="https://github.com/rancher/k3s/issues/1364" target="_blank">otherwise K3S will revert all changes and deploy Traefik the way it was</a>.</p>
<p>Finally, edit K3S configuration in <code>/etc/systemd/system/k3s.service</code> and add <code>&mdash;-no-deploy traefik</code></p>
<p>Verify that your ingress works correctly by creating an ingress for <a rel="noopener" href="https://hub.docker.com/_/hello-world" target="_blank">docker hello-world application</a>, making it available at <code>hello.&lt;yourdomain&gt;.com</code></p>
<pre><span>kind: Ingress<br>apiVersion: extensions/v1beta1<br>metadata:<br> name: hello<br>spec:<br> rules:<br> - host: hello.&lt;replaceme&gt;.me<br> http:<br> paths:<br> - backend:<br> serviceName: hello<br> servicePort: 80</span></pre>
<h2>Configuration Tips:</h2>
<p>If you wish to expose some HTTP service on your LAN, such as your router&rsquo;s dashboard, a NAS or some other device, y<a rel="noopener" href="https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-mapping-external-services" target="_blank">ou can create an endpoint and a corresponding service</a>, then use Ingress to direct HTTP traffic as usual. Please use TLS, authentication options in the ingress, and be careful exposing your router or anything else sensitive.</p>
<p>Cert manager issues and maintains up-to-date Let&rsquo;sEncrypt certificates for any ingress in your cluster. It is not strictly necessary, and you might have your own way of dealing with certificates.<br>
It&rsquo;s and is super-straight-forward to install:</p>
<pre><span><em># Kubernetes 1.15+</em><br>$ kubectl apply --validate<strong>=</strong>false -f <a rel="noopener" href="https://github.com/jetstack/cert-manager/releases/download/v0.14.3/cert-manager.crds.yaml" target="_blank">https://github.com/jetstack/cert-manager/releases/download/v0.14.3/cert-manager.crds.yaml</a><br>kubectl create namespace cert-manager<br># Helm 3<br>$ helm repo add jetstack <a rel="noopener" href="https://charts.jetstack.io" target="_blank">https://charts.jetstack.io</a><br>$ helm repo update<br>$ helm install cert-manager jetstack/cert-manager --namespace cert-manager --version v0.14.3</span></pre>
<p>In addition to installing helm, we need to configure Let&rsquo;s Encrypt Cluster Issuer, just apply the following yaml:</p>
<pre><span>apiVersion: cert-manager.io/v1alpha2<br>kind: ClusterIssuer<br>metadata:<br> name: letsencrypt<br>spec:<br> acme:<br> # The ACME server URL<br> server: <a rel="noopener" href="https://acme-v02.api.letsencrypt.org/directory" target="_blank">https://acme-v02.api.letsencrypt.org/directory</a><br> # Email address used for ACME registration<br> email: <a rel="noopener" href="mailto:vladimir@akopyan.me" target="_blank">replace@me.</a><br> # Name of a secret used to store the ACME account private key<br> privateKeySecretRef:<br> name: letsencrypt<br> # Enable the HTTP-01 challenge provider<br> solvers:<br> - http01:<br> ingress:<br> class: traefik / nginx</span></pre>
<p>Don&rsquo;t forget to replace ingress class with appropriate one for your cluster! Validate your setup by updating your ingress with TLS settings and an annotation that informs cert manager that it should create a certificate:</p>
<pre><span>kind: Ingress<br>apiVersion: extensions/v1beta1<br>metadata:<br> name: hello<br> annotations:<br> cert-manager.io/cluster-issuer: letsencrypt<br> ingress.kubernetes.io/ssl-redirect: 'true'<br>spec:<br> tls:<br> - hosts:<br> - hello.&lt;replaceme&gt;.me<br> secretName: hello.&lt;replaceme&gt;.me<br> rules:<br> - host: hello.&lt;replaceme&gt;.me<br> http:<br> paths:<br> - backend:<br> serviceName: hello<br> servicePort: 80</span></pre>
<p>You should see a pod appear with acme in it&rsquo;s name &mdash; it&rsquo;s responsible for responding to Let&rsquo;s Enrcypt acme challenge. Also, a secret will be created, and it will contain <code>tls.crt</code> and <code>tls.key</code> records. The key record will only be populated once the challenge completes &mdash; validate that it works. If you can monitor progress of a certificate being issues with <code>kubectl describe certs</code> and debug issues by checking logs of the cert manager pod.</p>
<p>Some applications aren&rsquo;t stateless: these are databases, image galleries, Wordpress, you name it. There are two ways of dealing with storage in Kubernetes &mdash; the plebian way and the proper way.</p>
<p>The plebian option is to directly expose a disk or directory from our server to the container &mdash; that&rsquo;s HostPath and <a rel="noopener" href="https://kubernetes.io/docs/concepts/storage/volumes/#local" target="_blank">Local Persistent Storage</a>. Hostpath is a total hack, the kubernetes scheduler could move the pod to a different machine at any time, and the data will not travel with it. The scheduler does respect Local PS and won&rsquo;d move the pod &mdash; it&rsquo;s a reasonable option if you are deploying a distributed database, or similar system which is designed to handle redundancy, replication, and clustering.</p>
<p>Distributed storage systems are designed to solve this problem, they pool together the storage space of all servers, and will provision a persistant volume for any pod that requests it. Data will be replicated to protect against disk failures, and it will move with the pod to a new node.</p>
<p><strong>Notable Open-source Projects:</strong></p>
<ul>
<li>
<a rel="noopener" href="https://ceph.io/" target="_blank">Ceph</a> &mdash; Block, Object and Network Attached storage. This battle-tested project significantly predates kubernetes, and can be used stanalone without K8S to create storage systems &mdash; it underlies block storage by <a rel="noopener" href="https://blog.digitalocean.com/why-we-chose-ceph-to-build-block-storage/" target="_blank">Digial Ocean</a>. It can be deployed on top of kubernetes with Rook.io. Requires entire disk or partition, which it will use raw &mdash; i.e. without a file system. Setup is not trivial.</li>
<li>
<a rel="noopener" href="https://github.com/Nexenta/edgefs" target="_blank">EdgeFs</a> &mdash; Block, Object and Network Attached storage. Cloud native storage tailored for kubernetes, supposedly production-ready. Beautiful dashboard, but the setup is very involved. Can use both folders and raw devices, but really the folder approach is impractical.</li>
<li>
<a rel="noopener" href="https://openebs.io/" target="_blank">OpenEBS</a> &mdash; Block Storage Only, requires whole device / partition. Setup is reasonably straightforward, this project has probably the largest amount of community support. It is very feature-rich, comes with prometheus metrics, etc.</li>
<li>
<a rel="noopener" href="https://github.com/longhorn/longhorn" target="_blank">Rancher Longhorn</a> &mdash; Block Storage Only, and by far the easiest distributed storage to deploy &mdash; you will be done in 5 minutes. This is what I ended up using. Installation steps:</li>
</ul>
<pre><span>git clone https://github.com/longhorn/longhorn &amp;&amp; cd longorn<br>kubectl create namespace longhorn-system<br>#Helm 3<br>helm install longhorn ./longhorn/chart/ --namespace longhorn-system</span></pre>
<p>That&rsquo;s it! It comes with a great dashboard, edit the it&rsquo;s service to Loadbalancer and open it in a browser &mdash; you will be presented with a summary of your cluster:</p>
</div>
</div>
<div>
<div>
<ul>
<li>In the Nodes tab, edit every node and add all the disks. They have to be formatted and mounted &mdash; you add them as a filepath.</li>
<li>If you have different classes of disks, like SSD and HDD, use <a rel="noopener" href="https://github.com/longhorn/longhorn/wiki/Feature:-Storage-Tag" target="_blank">storage tags</a>, and then create two storage classes with different diskSelectors:</li>
</ul>
<pre><span>apiVersion: storage.k8s.io/v1<br>kind: StorageClass<br>metadata:<br> annotations:<br> storageclass.beta.kubernetes.io/is-default-class: "false"<br> name: longhorn-hdd<br>parameters:<br> numberOfReplicas: "2"<br> staleReplicaTimeout: "30"<br> diskSelector: "hdd"<br>provisioner: driver.longhorn.io<br>reclaimPolicy: Delete<br>volumeBindingMode: Immediate</span></pre>
<ul>
<li>Longhorn only provides block storage, which can be attached to a single pod at a time. If you need NFS-style shared storage, you will have to standup a separate service in a container, on top of it. Same goes for object storage.</li>
<li>The UI has no authentication mechanism and allows anyone to delete all of your data . Once you&rsquo;ve configured longhorn, I would advice reverting the servicetype back to &lsquo;ClusterIP&rsquo;, and configure ingress as an authenticating proxy, at least with basic authentication.</li>
<li>It&rsquo;s worthwhile setting up backups of your storage, longhorn can be provided with a S3-compatible or NFS-compatiable storage.</li>
<li>To validate that it&rsquo;s working, deploy WordPress helm chart- it&rsquo;s will deploy two PVs, one for itself and one for MariaDB</li>
<li>The longhorn project is still in beta, <a rel="noopener" href="https://github.com/longhorn/longhorn/wiki/Roadmap" target="_blank">consult the roadmap</a> if you&rsquo;d like more info.</li>
</ul>
<p>Now your cluster has all the essentials &mdash; you are basically your own cloud provider. You can spend more time improving your cluster and deploying prometheus, grafana, and other services, or you could jump straight in and host your blog, or whatever else you have on your mind.</p>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Medium https://ift.tt/3beLIUZ<br>
<br>
May 9, 2020 at 09:59PM
